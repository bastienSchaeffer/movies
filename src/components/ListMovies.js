import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const ListMovies = ({
  error,
  isLoading,
  movies,
  onClickLoadMovies,
}) => (
  <Fragment>
    {
      // Display the error
      !!error
      && (
        <div data-testid="error">
          <p data-testid="error-message">{error}</p>
        </div>
      )
    }
    {
      // Display the loading
      isLoading
      && (
        <div data-testid="is-loading">
          <p>...Loading</p>
        </div>
      )
    }
    {
      // Display the movies
      movies.length
        ? (
          <ul data-testid="list-movies">
            {
              movies.map(({ id, title }) => (
                <li key={id} data-testid="movie">
                  <span>{title}</span>
                </li>
              ))
            }
          </ul>
        )
        : ''
    }
    <button
      type="button"
      data-testid="load-movie-button"
      onClick={onClickLoadMovies}
    >
      Load Movies
    </button>
  </Fragment>
);


ListMovies.propTypes = {
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  onClickLoadMovies: PropTypes.func.isRequired,
  movies: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      year: PropTypes.string,
    }),
  ),
};

ListMovies.defaultProps = {
  error: null,
  movies: [],
};
export default ListMovies;
