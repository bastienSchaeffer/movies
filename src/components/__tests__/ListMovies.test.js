import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import ListMovies from '../ListMovies';

/*
  Keep all clean
*/
afterEach(() => {
  cleanup();
  // eslint-disable-next-line no-console
  console.error.mockClear();
});
// eslint-disable-next-line no-console
console.error = jest.fn();

const onClickLoadMovies = jest.fn();
const propsComponent = {
  error: null,
  isLoading: false,
  movies: [],
  onClickLoadMovies,
};

describe('ListMovies', () => {
  describe('render', () => {
    test('renders no loading, no error, no movies and a button', async () => {
      // Act
      const { container } = render(<ListMovies {...propsComponent} />);
      // Assert
      expect(container).toMatchSnapshot();
    });

    test('renders an error and a button ', async () => {
      // Arrange
      const propsWithError = {
        ...propsComponent,
        error: 'Failed to load',
      };
      // Act
      const { container } = render(<ListMovies {...propsWithError} />);
      // Assert
      expect(container).toMatchSnapshot();
    });

    test('renders a loading indicator ', async () => {
      // Arrange
      const propsWithLoading = {
        ...propsComponent,
        isLoading: true,
      };
      // Act
      const { container } = render(<ListMovies {...propsWithLoading} />);
      // Assert
      expect(container).toMatchSnapshot();
    });

    test('renders the movies ', async () => {
      // Arrange
      const propsWithMovies = {
        ...propsComponent,
        movies: [
          { id: 'movie_1', title: 'Title movie one' },
          { id: 'movie_2', title: 'Title movie two' },
        ],
      };
      // Act
      const { container } = render(<ListMovies {...propsWithMovies} />);
      // Assert
      expect(container).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    test('load movies button should call the props action', async () => {
      // Act
      const { getByTestId } = render(<ListMovies {...propsComponent} />);
      const loadMovieButton = getByTestId('load-movie-button');
      fireEvent.click(loadMovieButton);
      // Assert
      expect(onClickLoadMovies).toHaveBeenCalledTimes(1);
    });
  });
});
