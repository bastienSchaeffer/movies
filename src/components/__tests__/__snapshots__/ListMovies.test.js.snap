// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMovies render renders a loading indicator  1`] = `
<div>
  <div
    data-testid="is-loading"
  >
    <p>
      ...Loading
    </p>
  </div>
  
  <button
    data-testid="load-movie-button"
    type="button"
  >
    Load Movies
  </button>
</div>
`;

exports[`ListMovies render renders an error and a button  1`] = `
<div>
  <div
    data-testid="error"
  >
    <p
      data-testid="error-message"
    >
      Failed to load
    </p>
  </div>
  
  <button
    data-testid="load-movie-button"
    type="button"
  >
    Load Movies
  </button>
</div>
`;

exports[`ListMovies render renders no loading, no error, no movies and a button 1`] = `
<div>
  
  <button
    data-testid="load-movie-button"
    type="button"
  >
    Load Movies
  </button>
</div>
`;

exports[`ListMovies render renders the movies  1`] = `
<div>
  <ul
    data-testid="list-movies"
  >
    <li
      data-testid="movie"
    >
      <span>
        Title movie one
      </span>
    </li>
    <li
      data-testid="movie"
    >
      <span>
        Title movie two
      </span>
    </li>
  </ul>
  <button
    data-testid="load-movie-button"
    type="button"
  >
    Load Movies
  </button>
</div>
`;
